### PR TITLE
attachment progress/retry

### PR DESCRIPTION
**Pull request by michaelkirk**
PTAL @charlesmchen

### CODE REVIEW

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 1]
```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
...
+                                                                 handler:^(UIAlertAction *_Nonnull action) {
+                                                                     [self.shareViewDelegate shareViewWasCancelled];
+                                                                 }];
+    [progressAlert addAction:progressCancelAction];
+
+
+    // Adding a subview to the alert controller like this is a total hack.
+    // ...but it looks good and given how short a progress view is, and the
+    // alert controller doesn't change much from version to version.
+#ifdef DEBUG
```
Comment (michaelkirk) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: e.g. 
screen shot 2017-12-20 at 8 20 11 pm

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 2]
```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
...
+        // Congratulations! You survived to see another iOS release.
+        OWSFail(@"Make sure progress view still looks good increment this version canary.");
+    }
+#endif
+    [progressAlert.view addSubview:self.progressView];
+    [self.progressView autoPinWidthToSuperviewWithMargin:24];
+    [self.progressView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:progressAlert.view withOffset:4];
+
+    void (^presentRetryDialog)(NSError *error) = ^(NSError *error) {
+        [fromViewController
```
Comment (michaelkirk) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: screen shot 2017-12-20 at 8 54 55 pm

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 3]
```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
...
+                                                                 handler:^(UIAlertAction *_Nonnull action) {
+                                                                     [self.shareViewDelegate shareViewWasCancelled];
+                                                                 }];
+    [progressAlert addAction:progressCancelAction];
+
+
+    // Adding a subview to the alert controller like this is a total hack.
+    // ...but it looks good, and given how short a progress view is and how
+    // little the alert controller changes, I'm not super worried about it.
+#ifdef DEBUG
```
Comment (michaelkirk) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: screen shot 2017-12-20 at 8 29 46 pm

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 4]
```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
...
+    if (@available(iOS 12, *)) {
+        // Congratulations! You survived to see another iOS release.
```
Comment (charlesmchen) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: Can you add the prefix TODO: to this comment

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 5]
```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
...
+    [progressAlert.view addSubview:self.progressView];
+    [self.progressView autoPinWidthToSuperviewWithMargin:24];
+    [self.progressView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:progressAlert.view withOffset:4];
+
+    void (^presentRetryDialog)(NSError *error) = ^(NSError *error) {
+        [fromViewController
+            dismissViewControllerAnimated:YES
+                               completion:^(void) {
+                                   AssertIsOnMainThread();
+                                   NSString *failureTitle
```
Comment (charlesmchen) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: Break out into `showSendFailureAlert` method?

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 6]
```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
...
+    DDLogDebug(@"%@ upload progress.", self.logTag);
+    AssertIsOnMainThread();
+    OWSAssert(self.progressView);
+
+    if (!self.outgoingMessage) {
+        DDLogDebug(@"%@ Ignoring upload progress until there is an outgoing message.", self.logTag);
+        return;
+    }
+
+    NSString *attachmentRecordId = self.outgoingMessage.attachmentIds.firstObject;
```
Comment (charlesmchen) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: _Nullable

File: SignalMessaging/attachments/SharingThreadPickerViewController.m (hunk 4 above)
Comment (michaelkirk) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: Done.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m (hunk 6 above)
Comment (michaelkirk) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: done!

File: SignalMessaging/attachments/SharingThreadPickerViewController.m (hunk 5 above)
Comment (michaelkirk) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: done.

**Pull request merged**